retrievalmarket/server/gsunpaidretrieval.go

```diff
@@ -0,0 +1,475 @@
+			g.failTransfer(state, completeErr)
+			return
+		}
+
+		// Fire markets blocks completed event
+		state.mkts.Status = retrievalmarket.DealStatusBlocksComplete
+		g.publishMktsEvent(retrievalmarket.ProviderEventBlocksCompleted, *state.mkts)
+
+		// Include a markets protocol Completed message in the response
+		voucher := &retrievalmarket.DealResponse{
+			ID:     state.mkts.DealProposal.ID,
+			Status: retrievalmarket.DealStatusCompleted,
+		}
+
+		const isAccepted = true
+		const isPaused = false
+		respMsg, err := message.CompleteResponse(msg.TransferID(), isAccepted, isPaused, voucher.Type(), voucher)
+		if err != nil {
+			g.failTransfer(state, fmt.Errorf("getting complete response: %w", err))
+			return
+		}
+
+		// Send the other peer a message that the transfer has completed
+		if err := g.dtnet.SendMessage(g.ctx, p, respMsg); err != nil {
+			err := fmt.Errorf("failed to send completion message to requestor %s: %w", p, err)
+			g.failTransfer(state, err)
+			return
+		}
+
+		state.cs.status = datatransfer.Completed
+		g.publishDTEvent(datatransfer.Complete, "", state.cs)
+		// Fire markets blocks completed event
+		state.mkts.Status = retrievalmarket.DealStatusCompleted
+		g.publishMktsEvent(retrievalmarket.ProviderEventComplete, *state.mkts)
+
+		log.Infow("successfully sent completion message to requestor", "peer", p)
+	})
+}
+
+func (g *GraphsyncUnpaidRetrieval) RegisterRequestorCancelledListener(listener graphsync.OnRequestorCancelledListener) graphsync.UnregisterHookFunc {
+	return g.GraphExchange.RegisterRequestorCancelledListener(func(p peer.ID, request graphsync.RequestData) {
+		_, state, intercept := g.isRequestForActiveUnpaidRetrieval(p, request)
+		if !intercept {
+			listener(p, request)
+			return
+		}
+
+		state.cs.status = datatransfer.Cancelled
+		g.publishDTEvent(datatransfer.Cancel, "", state.cs)
+		state.mkts.Status = retrievalmarket.DealStatusCancelled
+		g.publishMktsEvent(retrievalmarket.ProviderEventCancelComplete, *state.mkts)
+
+		g.untrackTransfer(p, state.cs.transferID)
+	})
+}
+
+func (g *GraphsyncUnpaidRetrieval) RegisterBlockSentListener(listener graphsync.OnBlockSentListener) graphsync.UnregisterHookFunc {
+	return g.GraphExchange.RegisterBlockSentListener(func(p peer.ID, request graphsync.RequestData, block graphsync.BlockData) {
+		_, state, intercept := g.isRequestForActiveUnpaidRetrieval(p, request)
+		if !intercept {
+			listener(p, request, block)
+			return
+		}
+
+		// When a data transfer is restarted, the requester sends a list of CIDs
+		// that it already has. Graphsync calls the sent hook for all blocks even
+		// if they are in the list (meaning, they aren't actually sent over the
+		// wire). So here we check if the block was actually sent
+		// over the wire before firing the data sent event.
+		if block.BlockSizeOnWire() == 0 {
+			return
+		}
+
+		// Fire block sent event
+		state.cs.sent += block.BlockSizeOnWire()
+		g.publishDTEvent(datatransfer.DataSent, "", state.cs)
```

Comment:
DataSentProgress as well? I have no idea why there are two events -- someone else put that in, but probably best for compatibility.

Reply:
The only place these events are being listened for is in boost (so it can display them to the user) so I don't think we need to replicate every event.

Reply:
ah, totally makes sense.